### DISCUSSION TITLE
Training any of the ultralytics library model for OBB data

### DISCUSSION THREAD

**Discussion by kedardg** [ultralytics/ultralytics#8264] [Q&A] [answered]
I am training a model for very small parts in OBB format. While training the model, I observed that the v8 series of OBB models struggle with it. Recently, I discovered the yolov8x-p2, which works really well for cases where objects are very small. 
Even if pretrained weights for yolov8x-p2 or redtr might not be available for OBB, is it possible for me to train these models for OBB?
If yes, then what changes should I make to execute OBB training on models other than yolov8?

**Comment by glenn-jocher** [comment, marked as answer]
Hello! I'm glad to hear that you've found the yolov8x-p2 model to be effective for detecting very small objects. 😊

To train models like yolov8x-p2 or redtr for OBB (Oriented Bounding Box) data, you would typically need to modify the model's architecture to output oriented bounding boxes instead of axis-aligned ones. However, this can be quite complex as it involves changing the prediction heads and the loss functions to handle the angle predictions.

For YOLOv8, we currently support OBB with specific models designed for this task, such as `yolov8n-obb.pt`. If you want to train a different model architecture for OBB, you would need to implement these modifications yourself, which requires a good understanding of the model's internals.

Here's a high-level overview of the steps you might take:
1. Adjust the model definition to include an additional output for the angle of rotation.
2. Modify the loss function to correctly penalize errors in angle prediction.
3. Adapt the data loading process to supply the correct OBB annotations.

Since this is a non-trivial task, we recommend using the provided OBB models if possible. If you're set on using a different architecture, you might consider reaching out to the community for contributions or collaborations on this feature.

Best of luck with your model training! 🚀

**Comment by hsluash** [comment]
Hi, I'm trying to train a yolov8-obb model. The command i use gives syntax error:
yolo obb train data=/home/ash/Downloads/obb-yolov8.1/ultralytics-8.1.0/dataset/data.yaml model=yolov8n-obb.pt epochs=10 imgsz=640
Traceback (most recent call last):
  File "/usr/local/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/cfg/__init__.py", line 381, in entrypoint
    check_dict_alignment(full_args_dict, {a: ''})
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/cfg/__init__.py", line 202, in check_dict_alignment
    raise SyntaxError(string + CLI_HELP_MSG) from e
SyntaxError: 'obb' is not a valid YOLO argument. 

please let me know how can i train the OBB model for my dataset?

**Reply by glenn-jocher** [reply]
Hello! It looks like there's a small mix-up in the command you're using. To train an OBB model, you don't need to specify `obb` in the command. Instead, you should use the `detect` task followed by the `train` mode. Here's the correct command format:

```bash
yolo detect train data=/path/to/your/data.yaml model=/path/to/yolov8n-obb.pt epochs=10 imgsz=640
```

Make sure that your `data.yaml` and model paths are correct, and this should start the training process for your OBB dataset. If you encounter any further issues, feel free to reach out! 😊

**Reply by hsluash** [reply]
I'm now running this command on a docker and find some issue.
root@4a180c88ffec:/app/ultralytics-8.1.0# yolo detect train data=/app/ultralytics-8.1.0/dataset/LicensePlateRecognition.v4-resized640_aug3x-accurate.yolov8-obb/data.yaml model=/app/ultralytics-8.1.0/dataset/yolov8n-obb.pt epochs=10 imgsz=640
Traceback (most recent call last):
  File "/opt/conda/bin/yolo", line 33, in <module>
    sys.exit(load_entry_point('ultralytics', 'console_scripts', 'yolo')())
  File "/usr/src/ultralytics/ultralytics/cfg/__init__.py", line 423, in entrypoint
    model = YOLO(model, task=task)
  File "/usr/src/ultralytics/ultralytics/engine/model.py", line 94, in __init__
    self._load(model, task)
  File "/usr/src/ultralytics/ultralytics/engine/model.py", line 146, in _load
    self.model, self.ckpt = attempt_load_one_weight(weights)
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 628, in attempt_load_one_weight
    ckpt, weight = torch_safe_load(weight)  # load ckpt
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 567, in torch_safe_load
    return torch.load(file, map_location='cpu'), file  # load
  File "/opt/conda/lib/python3.10/site-packages/torch/serialization.py", line 1014, in load
    return _load(opened_zipfile,
  File "/opt/conda/lib/python3.10/site-packages/torch/serialization.py", line 1422, in _load
    result = unpickler.load()
  File "/opt/conda/lib/python3.10/site-packages/torch/serialization.py", line 1415, in find_class
    return super().find_class(mod_name, name)
AttributeError: Can't get attribute 'OBBModel' on <module 'ultralytics.nn.tasks' from '/usr/src/ultralytics/ultralytics/nn/tasks.py'>

could you please help me with this?

**Reply by glenn-jocher** [reply]
@hsluash hey there! It seems like there might be an issue with the model file you're trying to load. The error message indicates that the `OBBModel` attribute is not found, which could suggest a version mismatch or a corrupted file.

Could you please try the following steps:
1. Ensure that you're using the correct version of the Ultralytics YOLOv8 that supports OBB.
2. Verify that the `yolov8n-obb.pt` file is not corrupted and is the correct OBB model file.

If the issue persists, consider re-downloading the model file or updating your Ultralytics YOLOv8 installation. If you need further assistance, feel free to share more details or reach out to the Ultralytics team. Good luck! 🚀

**Comment by Quang-minh-0211** [comment]
Hello, first let me thank you for answering my question. My team is from VietNam and now we are researching AI. I use Yolov8 to detect and recognize some objects from images. But I don't understand how actually the Ultralytics library works in training a model. Can you help me explain how actually the library works

**Reply by pderrenger** [reply]
@Quang-minh-0211 hello! We're delighted to hear about your interest and work in AI research with YOLOv8. 😊 In essence, the Ultralytics library simplifies the process of training object detection models by wrapping complex processes into an intuitive API. Here's a high-level overview of how training works with the library:

1. **Data Preparation**: Arrange your data (images and annotations) according to the expected format, usually with separate folders for images and annotation files.

2. **Configuration**: Define your model and training settings in YAML configuration files. This includes specifying your dataset paths, model architecture (e.g., `yolov8n.yaml`), and any other training hyperparameters.

3. **Training**: Leveraging PyTorch at its core, the library takes care of model training by iterating over your dataset, performing forward and backward passes, optimizing model weights, and saving checkpoints.

4. **Evaluation**: Throughout training, the model's performance is evaluated on a validation set to ensure it's learning correctly, giving insights through metrics like mAP (mean Average Precision).

5. **Exporting**: After training, you can export your model to different formats for deployment.

Here's a basic example of how you might start training using the CLI:

```bash
yolo train data=your_dataset.yaml model=yolov8n.yaml epochs=100
```

This commands the library to train a model specified by `yolov8n.yaml` on the dataset described in `your_dataset.yaml` for 100 epochs.

The strength of the library lies in its ability to abstract detailed implementation, allowing researchers and developers to focus more on experimenting with different settings and architectures rather than dealing with the underlying mechanics. I hope this helps clarify how the Ultralytics library facilitates training YOLOv8 models. If you have any more questions or need further assistance, feel free to ask! 🚀
